src/wat-parser.h

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2022 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef wasm_wat_parser_h
+#define wasm_wat_parser_h
+
+#include <optional>
+#include <string_view>
+
+#include "wasm.h"
+
+namespace wasm::WATParser {
+
+struct Ok {};
+
+struct None {};
+
+struct Err {
+  std::string msg;
+};
+
+template<typename T = Ok> struct Result {
+  std::variant<T, Err> val;
+
+  Result(Result<T>& other) = default;
+  Result(const Err& e) : val(std::in_place_type<Err>, e) {}
+  Result(Err&& e) : val(std::in_place_type<Err>, e) {}
+  template<typename U = T>
+  Result(U&& u) : val(std::in_place_type<T>, std::forward<U>(u)) {}
+
+  Err* getErr() { return std::get_if<Err>(&val); }
+  T& operator*() { return *std::get_if<T>(&val); }
+  T* operator->() { return std::get_if<T>(&val); }
+};
+
+template<typename T = Ok> struct MaybeResult {
+  std::variant<T, None, Err> val;
+
+  MaybeResult() : val(None{}) {}
+  MaybeResult(MaybeResult<T>& other) = default;
+  MaybeResult(const Err& e) : val(std::in_place_type<Err>, e) {}
+  MaybeResult(Err&& e) : val(std::in_place_type<Err>, std::move(e)) {}
+  template<typename U = T>
+  MaybeResult(U&& u) : val(std::in_place_type<T>, std::forward<U>(u)) {}
+
+  // Whether we have an error or a value. Useful for assignment in loops and if
+  // conditions where errors should not get lost.
+  operator bool() const { return !std::holds_alternative<None>(val); }
+
+  Err* getErr() { return std::get_if<Err>(&val); }
+  T& operator*() { return *std::get_if<T>(&val); }
+  T* operator->() { return std::get_if<T>(&val); }
+};
```

Comment:
> Basic C++ questions :sweat_smile::
> 
> * In which case we use `move` and when to use `forward`? Why does `Err` constructor use `move` and the generic constructor use `forward`?

IIUC, `std::forward` is like `std::move` when the value is an rvalue and otherwise it is like passing by reference. In other words it produces an rvalue reference when passed an rvalue reference and produces an lvalue reference when passed an lvalue reference. `std::move` is used when you know you are always getting an rvalue reference.

> * Why does `MaybeResult`'s `Err&&` constructor use `move` while `Result`'s `Err&&` constructor doesn't?

Ah, that looks like a mistake. They should both use `move`.

> * Why do we need an additional parameter `U` in addition to `T`? (I guess that's to use `forward`, which brings me back to the first question)

Not only does this allow us to handle lvalue references and rvalue references all in one case, it also improves ergonomics by allowing us to depend on converting constructors, so for example I can have `return Type::i32;` and have that `Type::BasicType` convert to `Type` and then convert to `Result<Type>`.

Reply:
> > > * In which case we use `move` and when to use `forward`? Why does `Err` constructor use `move` and the generic constructor use `forward`?
> > 
> > 
> > IIUC, `std::forward` is like `std::move` when the value is an rvalue and otherwise it is like passing by reference. In other words it produces an rvalue reference when passed an rvalue reference and produces an lvalue reference when passed an lvalue reference. `std::move` is used when you know you are always getting an rvalue reference.
> 
> So do we know we always get an rvalue reference for `Err` but not for other things?

Right!

> > > * Why do we need an additional parameter `U` in addition to `T`? (I guess that's to use `forward`, which brings me back to the first question)
> > 
> > 
> > Not only does this allow us to handle lvalue references and rvalue references all in one case, it also improves ergonomics by allowing us to depend on converting constructors, so for example I can have `return Type::i32;` and have that `Type::BasicType` convert to `Type` and then convert to `Result<Type>`.
> 
> Can you elaborate on how this works? Not sure what role each of the template parameters play when there's a single template param in usage...

`Type::i32` has type `Type::BasicType` and the compiler knows that I'm trying to return a `Result<Type>`, so it looks for a way to construct a `Result<Type>` from a `Type::BasicType`. The only candidate constructor is this constructor template with `U = Type::BasicType`. If we didn't have `U` and we used `Result(T&& t) ...` instead, this constructor would not be a candidate because `Type::BasicType` is not a `T` (i.e. it's not `Type`). You could imagine the compiler getting smarter and realizing that it can go `Type::BasicType -> Type -> Result<Type>`, but the language doesn't allow for that because that would introduce a bunch of other problems such as having many choices of paths from one type to another.

Reply:
> > > > * In which case we use `move` and when to use `forward`? Why does `Err` constructor use `move` and the generic constructor use `forward`?
> > > 
> > > 
> > > IIUC, `std::forward` is like `std::move` when the value is an rvalue and otherwise it is like passing by reference. In other words it produces an rvalue reference when passed an rvalue reference and produces an lvalue reference when passed an lvalue reference. `std::move` is used when you know you are always getting an rvalue reference.
> > 
> > 
> > So do we know we always get an rvalue reference for `Err` but not for other things?
> 
> Right!

I was asking more about... How do we know that? 

> > > > * Why do we need an additional parameter `U` in addition to `T`? (I guess that's to use `forward`, which brings me back to the first question)
> > > 
> > > 
> > > Not only does this allow us to handle lvalue references and rvalue references all in one case, it also improves ergonomics by allowing us to depend on converting constructors, so for example I can have `return Type::i32;` and have that `Type::BasicType` convert to `Type` and then convert to `Result<Type>`.
> > 
> > 
> > Can you elaborate on how this works? Not sure what role each of the template parameters play when there's a single template param in usage...
> 
> `Type::i32` has type `Type::BasicType` and the compiler knows that I'm trying to return a `Result<Type>`, so it looks for a way to construct a `Result<Type>` from a `Type::BasicType`. The only candidate constructor is this constructor template with `U = Type::BasicType`. If we didn't have `U` and we used `Result(T&& t) ...` instead, this constructor would not be a candidate because `Type::BasicType` is not a `T` (i.e. it's not `Type`). You could imagine the compiler getting smarter and realizing that it can go `Type::BasicType -> Type -> Result<Type>`, but the language doesn't allow for that because that would introduce a bunch of other problems such as having many choices of paths from one type to another.

I see. Thanks!